core/test/mpi/distributed/vector.cpp

```diff
@@ -0,0 +1,864 @@
+    auto vec = TestFixture::dist_vec_type::create(this->ref, this->comm);
+    auto rank = this->comm.rank();
+
+    vec->read_distributed(this->md, part.get());
+
+    GKO_ASSERT_EQUAL_DIMENSIONS(vec->get_size(), gko::dim<2>(6, 2));
+    if (rank == 1) {
+        GKO_ASSERT_EQUAL_DIMENSIONS(vec->get_local()->get_size(),
+                                    gko::dim<2>(6, 2));
+        GKO_ASSERT_MTX_NEAR(
+            vec->get_local(),
+            l({{0., 1.}, {2., 3.}, {4., 5.}, {6., 7.}, {8., 9.}, {10., 11.}}),
+            r<value_type>::value);
+    } else {
+        GKO_ASSERT_EQUAL_DIMENSIONS(vec->get_local()->get_size(),
+                                    gko::dim<2>(0, 2));
+    }
+}
+
+
+TYPED_TEST(VectorCreation, CanReadGlobalDeviceMatrixData)
+{
+    using it = typename TestFixture::global_index_type;
+    using d_md_type = typename TestFixture::d_md_type;
+    using part_type = typename TestFixture::part_type;
+    using vt = typename TestFixture::value_type;
+    d_md_type md{
+        this->ref, gko::dim<2>{6, 2},
+        gko::Array<it>{this->ref, I<it>{0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5}},
+        gko::Array<it>{this->ref, I<it>{0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1}},
+        gko::Array<vt>{this->ref, I<vt>{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}}};
+    auto part = gko::share(
+        part_type::build_from_contiguous(this->ref, {this->ref, {0, 2, 4, 6}}));
+    auto vec = TestFixture::dist_vec_type::create(this->ref, this->comm);
+    auto rank = this->comm.rank();
+    I<I<vt>> ref_data[3] = {
+        {{0, 1}, {2, 3}},
+        {{4, 5}, {6, 7}},
+        {{8, 9}, {10, 11}},
+    };
+
+    vec->read_distributed(md, part.get());
+
+    GKO_ASSERT_EQUAL_DIMENSIONS(vec->get_size(), gko::dim<2>(6, 2));
+    GKO_ASSERT_EQUAL_DIMENSIONS(vec->get_local()->get_size(),
+                                gko::dim<2>(2, 2));
+    GKO_ASSERT_MTX_NEAR(vec->get_local(), ref_data[rank], r<vt>::value);
+}
+
+
+TYPED_TEST(VectorCreation, CanReadGlobalMatrixDataScattered)
+{
+    using md_type = typename TestFixture::md_type;
+    using part_type = typename TestFixture::part_type;
+    using value_type = typename TestFixture::value_type;
+    md_type md{{0, 1}, {2, 3}, {4, 5}, {6, 7}, {8, 9}, {10, 11}};
+    auto part = gko::share(part_type::build_from_mapping(
+        this->ref, {this->ref, {0, 1, 2, 0, 2, 0}}, 3));
+    auto vec = TestFixture::dist_vec_type::create(this->ref, this->comm);
+    auto rank = this->comm.rank();
+    gko::dim<2> ref_size[3] = {{3, 2}, {1, 2}, {2, 2}};
+    I<I<value_type>> ref_data[3] = {
+        {{0, 1}, {6, 7}, {10, 11}},
+        {{2, 3}},
+        {{4, 5}, {8, 9}},
+    };
+
+    vec->read_distributed(md, part.get());
+
+    GKO_ASSERT_EQUAL_DIMENSIONS(vec->get_size(), gko::dim<2>(6, 2));
+    GKO_ASSERT_EQUAL_DIMENSIONS(vec->get_local()->get_size(), ref_size[rank]);
+    GKO_ASSERT_MTX_NEAR(vec->get_local(), ref_data[rank], r<value_type>::value);
```

Comment:
```suggestion
    GKO_ASSERT_MTX_NEAR(vec->get_local(), ref_data[rank], 0.0);
```
also for other test which only do data movement/assignment